### PR TITLE
[Sema] Diagnose unsound pointer conversions

### PR DESCRIPTION
**Pull request by hamishknight**
*(This is an updated version of https://github.com/apple/swift/pull/20467)*

In this PR is the logic to emit a diagnostic for unsound array-to-pointer, string-to-pointer, and inout-to-pointer conversions. This is done through the introduction of an underscored attribute `@_nonEphemeral` that denotes a pointer parameter which cannot accept pointer conversions that are only valid for the duration of the call they're passed to.

This includes all array-to-pointer and string-to-pointer conversions, and most inout-to-pointer conversions with the exception of:

- Global and static variables
- Stored properties on such variables that are accessed directly (e.g excluding ones with property observers, ones on classes, protocols, and resilient types in general)
- Force unwraps of any of the above

Various pointer parameters in the stdlib have been marked as `@_nonEphemeral`, primarily pointer parameters for initialisers on pointer types. In addition, this PR infers `@_nonEphemeral` on pointer parameters for memberwise initialisers and enum case constructors.

By default, the diagnostic emitted will be warning. I have added a frontend flag to allow upgrading to an error, and hopefully this will allow us to transition to an error by default in a future language version.

Resolves [SR-2790](https://bugs.swift.org/browse/SR-2790).

### CODE REVIEW

**Comment by hamishknight** [comment]
Given how large this PR is, I've split off the commits that can be landed separately with no intended functionality change: #27696. Once that has been landed, I can further peel off the main `@_nonEphemeral` implementation, leaving this PR with just the stdlib + `@_nonEphemeral` inference changes (last 2 commits).

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
I've peeled off the `@_nonEphemeral` implementation into https://github.com/apple/swift/pull/27705.

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by rjmccall** [comment]
FWIW, this is essentially an escape analysis and `@_nonEphemeral` is essentially `@escaping` applied to pointers.

**Comment by hamishknight** [comment]
@rjmccall Yup, I wanted to keep this distinct from `@escaping` for a couple of reasons:

- Given this isn't an official feature, I wanted to make sure the attribute was non-user-facing, so we don't expose it in things like code completion or allow the user to start marking their pointer parameters as `@escaping`. I also wanted to make sure this attribute doesn't get considered in things like name mangling so it doesn't affect ABI. 

- `@_nonEphemeral` can currently be used to mark other kinds of pointer parameters that don't escape the pointer passed, but still shouldn't accept temporary pointer conversions. For example, one such parameter in this PR is:

  ```swift
  moveInitialize(@_nonEphemeral from source: UnsafeMutablePointer, count: Int)
  ```

  as passing a temporary pointer conversion here could lead to over releasing the pointee.

- As you say, I don't think we intend to build a fully safe pointer model. In particular, applying the same restrictions as non-escaping function parameters to pointer parameter would almost certainly be too source breaking.

**Review by atrick — APPROVED** [review]
This is immensely exciting to me! These checks will definitely catch a massive amount of nasty C interop bugs. By combining these sort of diagnostics with some Swift-for-C-Programmer documentation we can resolve some major pain points. I hope we can turn the diagnostics into errors in the release after next. Should we have a bug for that?

@rjmccall mentioned an escape analysis. I don't see any such thing in this PR, except that the manual additional of `@nonEphemeral` could be subsumed by a more general analysis. Eventually we should handle general escaping pointers, such as rdar://problem/25001764 Detect when a temporary pointer to a writeback buffer is persisted after the function returns:
```
func takesPtr(_ ptr: UnsafePointer<UInt64>) {}

private func _ptr(_ ptr: UnsafePointer<UInt64>) -> UnsafePointer<UInt64> {
  return UnsafePointer<UInt64>(ptr)
}

var x = UInt64(0)
takesPtr(_ptr(&x))
```

At that level, the diagnostic should be a SIL pass. We've been moving diagnostics that depend on dataflow analysis out of the type checker and into SIL passes where they can be expressed more robustly and without adding type checker constraints. I suspect the thing to do at that point would be to layer the diagnostics, so most of the logic in this PR stays in the type checker with more difficult cases handled later. At the SIL level, I don't think we need the `@nonEphemeral` attribute, since the job of the analysis is to infer that.

I am curious though... is it necessary for the precision of these diagnostics to be part of the constraint system? I don't see why it would be. In fact, it I seems to me that we do *not* want pointer lifetime mistakes to affect overload resolution. If I'm missing something, please clarify.

I really don't understand all the implications of adding constraints to the type checker, but overall your implementation looks extremely thorough and high quality.

**Comment by hamishknight** [comment]
@atrick Thanks for the review! It would be great to see this combined with a SIL pass to diagnose more tricky cases, and I'd certainly be happy to look into that as a potential follow-up :)

> I hope we can turn the diagnostics into errors in the release after next. Should we have a bug for that?

Sure thing, I'll file one after this gets merged.

> I am curious though... is it necessary for the precision of these diagnostics to be part of the constraint system? I don't see why it would be. In fact, it I seems to me that we do not want pointer lifetime mistakes to affect overload resolution. If I'm missing something, please clarify.

It's not entirely necessary for this to be a part of the constraint system, though it does allow us to emit better diagnostics for things like ambiguities, where we can inform the user that none of the candidates are viable because they don't accept temporary pointer conversions.

It will also allow for composition with other constraint system diagnostics, so we can tell the user earlier that what they're trying to do is invalid. Unfortunately this doesn't work so well at the moment as we discard additional fixes with the same locator, but eventually it would be nice to emit a temporary pointer error on the following along with an optional unwrap error:

```swift
func foo() {
  var x: Int?
  let ptr = UnsafePointer<Int>(&x)
}
```

I do agree that ideally `@_nonEphemeral` probably shouldn't affect overload resolution, though I don't think this should be a major issue in practice. One slight benefit is that it can help uncover declarations that should have been marked `@_nonEphemeral` in the first place.

**Comment by hamishknight** [comment]
If @xedin and @jrose-apple are also happy to review this as a whole, then I'll close https://github.com/apple/swift/pull/27705.

File: include/swift/AST/Attr.def
Comment (jrose-apple) [inline include/swift/AST/Attr.def]: Not exactly APIStableToAdd, since it's going to produce an error.

File: include/swift/AST/DiagnosticsSema.def
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: I don't know if it's a good idea to put "currently" in a diagnostic if we have no concrete plans to do anything else.

File: include/swift/AST/DiagnosticsSema.def
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: This is okay but it'd be nice to get the names in here too (see `argument_out_of_order_named_named`). That can be follow-up work for sure, though.

File: include/swift/Option/FrontendOptions.td
Comment (jrose-apple) [inline include/swift/Option/FrontendOptions.td]: It might be nice™ to phrase this as `-enable…` so that you can add a `-disable…` as well.

File: lib/Sema/CSFix.cpp
Comment (jrose-apple) [inline lib/Sema/CSFix.cpp]: If you're going to do this, it's going to copy anyway. Better to construct the `std::string` in place.

File: stdlib/public/core/UnsafePointer.swift
Comment (jrose-apple) [inline stdlib/public/core/UnsafePointer.swift]: This and `moveAssign` can be ephemeral; the pointer argument isn't persisted.

**Review by jrose-apple — COMMENTED** [review]
The non-ConstraintSystem parts seem good, but I'm also pretty unhappy with this being part of overload resolution. It feels like something that MiscDiagnostics or even SIL diagnostics should handle instead—after we've type-checked everything, find out if we relied on an ephemeral pointer.

If you want to keep going with this approach, @xedin (or @hborla) should review the ConstraintSystem parts.

File: include/swift/AST/Attr.def
Comment (hamishknight) [reply include/swift/AST/Attr.def]: @jrose-apple I was planning on changing it once we errored by default, I guess there's no harm in doing it now though, I'll go ahead and make that change.

File: include/swift/AST/DiagnosticsSema.def
Comment (hamishknight) [reply include/swift/AST/DiagnosticsSema.def]: @jrose-apple Good idea! I'll do this as a follow-up, as we'll probably want to add a new "get argument label" API to `ArgumentMismatchFailure` and `FunctionArgApplyInfo`.

File: include/swift/Option/FrontendOptions.td
Comment (hamishknight) [reply include/swift/Option/FrontendOptions.td]: @jrose-apple Hmm, I only really intended for this to be a temporary flag for testing the error behaviour, and was planning on removing it once we error-by-default. Do you think it would be worth keeping an option to downgrade to a warning?

File: stdlib/public/core/UnsafePointer.swift
Comment (hamishknight) [reply stdlib/public/core/UnsafePointer.swift]: @jrose-apple While the pointer argument isn't persisted, passing a temporary pointer conversion is still unsound because we'll over-release the value. We could emit a more specialized diagnostic for this, but I didn't think it was worth implementing in this PR. What do you think?

**Comment by hamishknight** [comment]
@jrose-apple Thanks for the review! I'll go ahead and make those changes. Note that @xedin has reviewed the constraint system parts of this over on https://github.com/apple/swift/pull/27705.

Regarding the fact that `@_nonEphemeral` affects overload resolution, I guess one simple solution would to change the `TreatEphemeralAsNonEphemeral` fix such that it doesn't affect the `SK_Fix` score of the solution (like warning fixes currently do, but we'd then emit an error). This would allow us to benefit from composition with other constraint system diagnostics without affecting overload resolution. Does this sound reasonable @xedin?

File: include/swift/Option/FrontendOptions.td
Comment (jrose-apple) [reply include/swift/Option/FrontendOptions.td]: It might help with the initial adoption within Apple, even if we eventually take the flag away.

File: stdlib/public/core/UnsafePointer.swift
Comment (jrose-apple) [reply stdlib/public/core/UnsafePointer.swift]: Hm, okay, I didn't think that through and that's evidence that it's best to leave it. People can explicitly use `withUnsafeMutableBufferPointer` or whatever if they want to get around this.

**Comment by hamishknight** [comment]
Thanks again everyone for the review! I've gone ahead and changed things such the diagnostic no longer affects overload resolution. I've also rebased to resolve a bunch of merge conflicts.

If possible I'd like to keep this logic a part of the constraint system, as it's much easier to detect and attach a fix there when evaluating the various X-to-pointer conversions rather than attempting to traverse the AST in MiscDiagnostics.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp]: Is `impact > 0` used here to avoid `worseThanBestSolution()` check? This condition seems strange because impact of `0` doesn't change anything which means that the solution was already `worse`, so instead of failing right away we'd just fail a bit later after simplifying  one of the active constraints...

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: @xedin Ah okay, I was under the impression that because we're not touching the score, there wouldn't be much point in checking `worseThanBestSolution`, but I guess there are other cases where we're touching the score and not checking it. Should we also be checking `worseThanBestSolution` for when the fix is a warning then?

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: I think we should just ways do that and reset `impact` for warning to be `0` because I believe it was always `1` before that's why that check exists.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Or rather call the `increaseScore` only if it's not a warning maybe.

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: @xedin Yeah, I was thinking we could just move the call to `worseThanBestSolution` to just below this if statement. Checking `impact > 0` is still useful as it'll avoid us printing the `(increasing score due to ` in the diagnostics dump (though we could move the check into `increaseScore`).

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: I think we should just move that check to the `increaseScore` instead, seems like a better place for it.

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: @xedin Thanks, I've gone ahead and made that change. Does it look good?

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Looks great, thank you!

**Comment by jrose-apple** [comment]
I still think it doesn't make sense for this to be in the constraint system at all, but I think I'm outvoted.

**Comment by xedin** [comment]
I can't talk about SIL diagnostics but I'm hopping that we can remove MiscDiagnostics and instead make that logic be a part of constraint system represented by fixes. Current behavior where solver would produce a solution which can't either be applied to an AST or lowered doesn't really make much sense to me...

I mentioned in one of my previous comments that I don't really like the idea of adding new attribute easier and we ought to detect this somehow differently maybe in SIL diagnostics?

**Comment by hamishknight** [comment]
It may be possible to re-implement this as a SIL diagnostic, and I'd be happy to look into doing that while looking into @atrick's idea of implementing a data flow diagnostic to catch more cases of unsound temporary pointer escapes. I'm not sure we'd be able to completely do away with the attribute though for cross-module cases where we can't see the implementation of the callee so don't know if they're immediately escaping the passed pointer or not.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test OS X platform

**Comment by hamishknight** [comment]
@swift-ci please test OS X platform

**Comment by hamishknight** [comment]
@swift-ci please test OS X platform

**Comment by jrose-apple** [comment]
The thing I like about MiscDiagnostics is that it's *not* mixed up with deciding how a program behaves. The ConstraintSystem decides what the user said, and MiscDiagnostics figures out whether it's actually what they meant or actually allowed. You could sink all of that down to SIL, I guess, but I'm not sure what benefit you'd get.

In my head, MiscDiagnostics would ideally be a pluggable visitor the way SIL optimizer passes are pluggable; if you have an expression-level diagnostic, you just add it to the visitor. Then the visitor does a single walk of the AST. (I might be biased by my static analyzer past; the analyzer's "checkers" work a lot like this.)

**Comment by atrick** [comment]
The purpose of sinking the diagnostic into SIL is to leverage being able to infer `@nonEphemeral` via an escape analysis of the pointer value.

**Comment by jrose-apple** [comment]
Hamish is right: non-ephemerality is a property of the function even as the body changes, so it can't be inferred from the body. (In particular, if a function happens to not escape a pointer today, it should still be able to reserve the privilege to do so.)

**Comment by atrick** [comment]
When `@nonEphemeral` *is* used as an API annotation I don't see an advantage of diagnosing it in SIL. I do like the idea that the diagnostic provably cannot affecting type resolution in any other way.

**Comment by hamishknight** [comment]
Just rebased to resolve a merge conflict. @xedin @atrick if it's okay with you, I'd like to get this merged sometime Monday so we can get this into 5.2. I'm more than happy to make any further changes and cherry-pick to 5.2 if needed.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test Linux platform

**Comment by hamishknight** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@hamishknight I'm fine with this, go ahead and merge.

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
I checked, and @atrick is also happy for this to be merged.

**Comment by hamishknight** [comment]
Thanks everyone! I'll work on adding argument labels to the diagnostics, as well as a changelog entry as follow-ups.

**Pull request merged**
